subprojects/base-services/src/main/java/org/gradle/internal/Actions.java

```diff
@@ -221,4 +223,84 @@ public void execute(T t) {
         };
     }
 
+    /**
+     * Composes 2 actions with set semantics. The order of parameters is important
+     * because it will mutate and return the first argument if it is already an
+     * {@link ActionSet<T> action set}. This is in particular useful to avoid eager
+     * initialization of sets, or actions, when in most of the cases the action is
+     * absent or single. If the first parameter is null, then it will return the second
+     * parameter, allowing fluent initialization of a set.
+     *
+     * @param a first action, maybe null
+     * @param b second action, must not be null
+     * @param <T> the type of the action subject
+     * @return an action representing the set of (a,b)
+     */
+    @SuppressWarnings("unchecked")
+    public static <T> Action<T> set(Action<T> a, Action<T> b) {
```

Comment:
Conditionally mutating the first argument is kind of odd semantics.

Possibly we have a mutable `ComposableAction` to which you can add things and whose implementation can optimise for zero or a small number of actions (which is the most common).

Or this method could return an immutable chaining action, that just runs A and then runs B, optimising for `Actions.doNothing()` and `A.equals(B)`.

Regardless, the implementation should deal with `Actions.doNothing()`.